File: src/runtimes/dotnetcore/src/ChildLangVisitor.cs
using System;
using System.Collections.Generic;
using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Tree;
using static LanguageBaseParser;

public class ChildLangVisitor : LanguageBaseBaseVisitor<object>
{
    private Dictionary<string, object> ENVIRONMENT = new Dictionary<string, object>();

    public override object VisitCommand_print([NotNull] Command_printContext context)
    {
        bool first = true;
        foreach (var item in context.children)
        {
            if (first)
            {
                first = false;
                continue;
            }

            Console.Write($"{Eval(item)} ");
        }
        Console.WriteLine();

        return base.VisitCommand_print(context);
    }

    public override object VisitCommand_read([NotNull] Command_readContext context)
    {
        var value = Console.ReadLine();
        ENVIRONMENT.Add(context.VARIABLE().ToString(), value);
        return base.VisitCommand_read(context);
    }

    public override object VisitCommand_assign([NotNull] Command_assignContext context)
    {
        var value = Eval(context.children[1]).ToString();
        ENVIRONMENT[context.children[0].ToString()] = value;

        return base.VisitCommand_assign(context);
    }

    public override object VisitCommand_block_func([NotNull] Command_block_funcContext context)
    {
        var body = (BlockContext)context.GetChild(3);
        ENVIRONMENT.Add(context.VARIABLE().ToString(), body);
        return null;
    }

    public override object VisitCommand_call([NotNull] Command_callContext context)
    {
        var body = (BlockContext)ENVIRONMENT[context.VARIABLE().ToString()];
        return VisitBlock(body);
    }

    public override object VisitCommand_block_if([NotNull] Command_block_ifContext context)
    {
        if (context.ChildCount < 3)
            throw new ChildLangRuntimeException();

        bool condition = Eval(context.GetChild<Bool_argContext>(0));
        if (condition)
        {
            var block = (BlockContext)context.GetChild(3);
            return Visit(block);
        }
        else if (context.ChildCount == 5)
        {
            var elseContext = (Command_block_elseContext)context.GetChild(4);
            if (elseContext != null)
            {
                return Visit(elseContext.block());
            }
        }

        return null;
    }

    public override object VisitCommand_block_while([NotNull] Command_block_whileContext context)
    {
        var conditionExp = (Bool_argContext)context.GetChild(1);
        var body = (BlockContext)context.GetChild(3);
        while (true)
        {
            bool condition = Eval(conditionExp);
            if (!condition)
                break;
            Visit(body);
        }
        return base.Visit(body);
    }

    private object Eval(IParseTree arg)
    {
        if (arg is TerminalNodeImpl terminalNode)
            return Eval(terminalNode);

        if (arg is Math_argContext mathArg)
            return Eval(mathArg);

        if (arg is Bool_argContext boolArg)
            return Eval(boolArg);

        if (arg is Command_argContext commandArg)
        {
            var args = new List<object>();
            foreach (var item in commandArg.children)
            {
                var subEval = Eval(item);
                args.Add(subEval);
            }

            return args.Count == 1 ? args[0] : args;
        }
        throw new ChildLangRuntimeException();
    }

    private object Eval(TerminalNodeImpl arg)
    {
        if (arg.Symbol.Type == STRING)
            return arg.ToString();

        if (arg.Symbol.Type == NUMBER)
            return arg.ToString();

        if (arg.Symbol.Type == VARIABLE)
            return ENVIRONMENT[arg.ToString()];

        if (arg.Symbol.Type == BOOL_TRUE)
            return true;

        if (arg.Symbol.Type == VARIABLE)
            return false;

        throw new ChildLangRuntimeException();
    }

    /// <summary>
    /// todo: refactor
    /// </summary>
    /// <param name="arg"></param>
    /// <returns></returns>
    private bool Eval(Bool_argContext arg)
    {
        if (arg.children.Count == 1 && arg.children[0] is TerminalNodeImpl node)
        {
            if (node.Symbol.Type == BOOL_TRUE)
                return true;

            if (node.Symbol.Type == BOOL_FALSE)
                return false;

            throw new ChildLangRuntimeException();
        }

        if (arg.ChildCount != 3)
            throw new ChildLangRuntimeException();

        string left = GetTerminalString(GetLastTerminalNodeChild(arg.GetChild(0), 0));
        var op = (TerminalNodeImpl)arg.children[1];
        string right = GetTerminalString(GetLastTerminalNodeChild(arg.GetChild(2), 0));

        if (op.Symbol.Type == BOOL_EQ)
            return left == right; // compare bool as string

        if (op.Symbol.Type == BOOL_GT)
            return double.Parse(left) > double.Parse(right);

        if (op.Symbol.Type == BOOL_LT)
            return double.Parse(left) < double.Parse(right);


        throw new ChildLangRuntimeException();
    }

    private double Eval(Math_argContext arg)
    {
        double value = 0;

        var left = GetDoubleValue(GetLastTerminalNodeChild(arg.GetChild(0), 0));
        var op = (TerminalNodeImpl)arg.children[1];
        var right = GetDoubleValue(GetLastTerminalNodeChild(arg.GetChild(2), 0));

        if (op.Symbol.Type == ADD)
            return left + right;

        if (op.Symbol.Type == SUB)
            return left - right;

        if (op.Symbol.Type == MUL)
            return left * right;

        if (op.Symbol.Type == DIV)
            return left / right;

        return value;
    }

    private TerminalNodeImpl GetLastTerminalNodeChild(IParseTree context, int index)
    {
        if (context.ChildCount == 0)
            return (TerminalNodeImpl)context;

        return GetLastTerminalNodeChild(context.GetChild(0), index);
    }

    private string GetTerminalString(TerminalNodeImpl t)
    {
        if (t.Symbol.Type == VARIABLE)
            return ENVIRONMENT[t.ToString()].ToString();

        if (t.Symbol.Type == NUMBER)
            return t.GetText();

        throw new ChildLangRuntimeException();
    }

    private double GetDoubleValue(TerminalNodeImpl t)
    {
        if (t.Symbol.Type == VARIABLE)
            return Convert.ToDouble(ENVIRONMENT[t.ToString()]);

        return double.Parse(t.GetText());
    }
}

File: src/runtimes/javascript/src/ChildLangVisitor.ts
import * as P from "./grammar/LanguageBaseParser";
import { LanguageBaseVisitor } from "./grammar/LanguageBaseVisitor";
import { TerminalNode } from "antlr4ts/tree/TerminalNode";
import { ParseTree } from "antlr4ts/tree/ParseTree";
import { LanguageBaseLexer } from "./grammar/LanguageBaseLexer";

import { UserInputOutput } from "./UserInputOutput";
import { ErrorNode } from "antlr4ts/tree/ErrorNode";
import { RuleNode } from "antlr4ts/tree/RuleNode";

export class ChildLangVisitor implements LanguageBaseVisitor<object>
{
    private defaultResult: Object = new Object();
    private ENVIRONMENT: Map<string, any> = new Map<string, any>();

    constructor(private io: UserInputOutput) { }

    visitCommand_print(ctx: P.Command_printContext) {
        for (var i = 1; i < ctx.childCount; i++) {
            const value = this.eval(ctx.getChild(i));
            let valueToWrite: string = "" + value;

            this.io.write(valueToWrite, false);

            if (i < ctx.childCount)
                this.io.write(' ', false);
        }
        this.io.write('', true);
        return this.defaultResult;
    }

    async visitCommand_read(ctx: P.Command_readContext) {
        var line = await this.io.readLine();
        this.ENVIRONMENT.set(ctx.VARIABLE().text, line);
        return this.defaultResult;
    }

    visitCommand_assign(ctx: P.Command_assignContext) {
        var value = this.eval(ctx.getChild(1));
        this.ENVIRONMENT.set(ctx.VARIABLE().text, value);
        return this.defaultResult;
    }

    async visitCommand_call(ctx: P.Command_callContext) {
        var block = this.ENVIRONMENT.get(ctx.VARIABLE().text);
        return await this.visit(block);
    }

    async visitCommand_block_if(ctx: P.Command_block_ifContext) {
        let condition: boolean = this.eval(ctx.bool_arg());
        if (condition) {
            return await this.visit(ctx.block());
        } else {
            var elseBlock = ctx.command_block_else();
            if (elseBlock != null) {
                return await this.visit(elseBlock.block());
            }
        }
        return this.defaultResult;
    }

    visitCommand_block_func(ctx: P.Command_block_funcContext) {
        this.ENVIRONMENT.set(ctx.VARIABLE().text, ctx.block());
        return this.defaultResult;
    }

    async visitCommand_block_while(ctx: P.Command_block_whileContext) {
        while (this.eval(ctx.bool_arg())) {
            await this.visit(ctx.block());
        }
        return this.defaultResult;
    }

    async visit(tree: ParseTree): Promise<object> {
        //this.log(tree.toStringTree());
        if (
            tree instanceof P.LangContext ||
            tree instanceof P.BlockContext ||
            tree instanceof P.CommandContext
        ) {
            if (tree.children != null) {
                for(let i = 0; i < tree.childCount; i++) {
                    const child = tree.getChild(i);
                    await this.visit(child);
                }
            }
        } else {
            if (tree instanceof P.Command_printContext)
                return this.visitCommand_print(tree);

            if (tree instanceof P.Command_readContext)
                return await this.visitCommand_read(tree);

            if (tree instanceof P.Command_assignContext)
                return this.visitCommand_assign(tree);

            if (tree instanceof P.Command_callContext)
                return await this.visitCommand_call(tree);

            if (tree instanceof P.Command_block_ifContext)
                return await this.visitCommand_block_if(tree);

            if (tree instanceof P.Command_block_funcContext)
                return this.visitCommand_block_func(tree);

            if (tree instanceof P.Command_block_whileContext)
                return await this.visitCommand_block_while(tree);
        }
        return this.defaultResult;
    }

    eval(tree: ParseTree): any {
        if (tree instanceof P.Command_argContext) {
            if (tree.childCount == 1) {
                return this.eval(tree.getChild(0));
            }
        } else if (tree instanceof TerminalNode) {
            if (tree.symbol.type == LanguageBaseLexer.STRING)
                return tree.text;

            if (tree.symbol.type == LanguageBaseLexer.NUMBER)
                return tree.text;

            if (tree.symbol.type == LanguageBaseLexer.BOOL_TRUE)
                return true; // TODO:

            if (tree.symbol.type == LanguageBaseLexer.BOOL_FALSE)
                return false; // TODO:

            if (tree.symbol.type == LanguageBaseLexer.VARIABLE) {
                var value = this.ENVIRONMENT.get(tree.text);
                if (value != null)
                    return value;
            }
        } else if (tree instanceof P.Math_argContext) {
            if (tree.childCount == 1) {
                return this.eval(tree.getChild(0));
            } else if (tree.childCount == 3) {
                const left = parseFloat(this.eval(tree.getChild(0)));
                var operator = tree.getChild(1) as TerminalNode;
                const right = parseFloat(this.eval(tree.getChild(2)));

                if (operator.symbol.type == LanguageBaseLexer.ADD)
                    return left + right;
                if (operator.symbol.type == LanguageBaseLexer.SUB)
                    return left - right;
                if (operator.symbol.type == LanguageBaseLexer.MUL)
                    return left * right;
                if (operator.symbol.type == LanguageBaseLexer.DIV)
                    return left / right;
            }
        } else if (tree instanceof P.Bool_argContext) {
            if (tree.childCount == 1) {
                return this.eval(tree.getChild(0));
            } else if (tree.childCount == 3) {
                const left = this.eval(tree.getChild(0));
                var operator = tree.getChild(1) as TerminalNode;
                const right = this.eval(tree.getChild(2));

                if (operator.symbol.type == LanguageBaseLexer.BOOL_EQ)
                    return (left == right);
                else if (operator.symbol.type == LanguageBaseLexer.BOOL_GT)
                    return (parseFloat(left) > parseFloat(right));
                else if (operator.symbol.type == LanguageBaseLexer.BOOL_LT)
                    return (parseFloat(left) < parseFloat(right));
            }
        }

        throw new Error("tree can not evaluated. " + tree.toStringTree());
    }

    visitChildren(node: RuleNode): object {
        throw new Error("Method not implemented.");
    }
    visitTerminal(node: TerminalNode): object {
        throw new Error("Method not implemented.");
    }
    visitErrorNode(node: ErrorNode): object {
        throw new Error("Method not implemented.");
    }

    log(message?: any, ...optionalParams: any[]): void {
        console.debug("[log", message, "]", optionalParams);
    }
}